src/util/generate_root.c

```diff
@@ -0,0 +1,73 @@
+/******************************************************************************
+ *
+ * Copyright 2018 Xaptum, Inc.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License
+ *
+ *****************************************************************************/
+#include <string.h>
+#include <xtt.h>
+
+int generate_root(struct cli_params* params)
+{
+    const char *privkey_filename = params->privkey;
+    const char *pubkey_filename = params->pubkey;
+    const char *id_filename = params->id;
+    char *root_pub_file = "root_pub.bin";
+    char *root_priv_file = "root_priv.bin";
+    xtt_ecdsap256_pub_key pub;
+    xtt_ecdsap256_priv_key priv;
+    xtt_certificate_root_id root_id;
+
+    unsigned char idpubkey[sizeof(xtt_ecdsap256_pub_key)+sizeof(xtt_certificate_root_id)];
+
+    // 0) Read root id from file, giving root_id a random number if there is no file given
+    if (id_filename != NULL) {
+        assert(-1 != read_from_file(id_filename, root_id.data, sizeof(xtt_certificate_root_id)));
+    } else
+    {
+        xtt_crypto_get_random(root_id.data, sizeof(xtt_certificate_root_id));
+    }
+
+    // 1) Generate root's keypair
+    if (0 != xtt_crypto_create_ecdsap256_key_pair(&pub, &priv)) {
+        fprintf(stderr, "Error creating ecdsap256 keypair\n");
+        return 1;
+    }
+
+    memcpy(idpubkey, root_id.data, sizeof(xtt_certificate_root_id));
+    memcpy(&idpubkey[sizeof(xtt_certificate_root_id)], pub.data, sizeof(xtt_ecdsap256_pub_key));
+
+    if (pubkey_filename != NULL && privkey_filename != NULL)
```

Comment:
This checking for NULL and setting of defaults should be done by the tool, not in this library function.